src/components/ColumnPicker/NewColumnPickerCategories.tsx

```diff
@@ -0,0 +1,120 @@
+import { Box, Button, List, Text } from "@saleor/macaw-ui/next";
+import React, { useMemo, useState } from "react";
+import { FormattedMessage } from "react-intl";
+
+import messages from "./messages";
+import { NewColumnPickerAvailableNodes } from "./NewColumnPickerAvailableNodes";
+import { NewColumnPickerPagination } from "./NewColumnPickerPagination";
+import { ColumnCategory } from "./useColumns";
+import { getExitIcon, getExitOnClick } from "./utils";
+
+export interface NewColumnPickerCategoriesProps {
+  columnCategories: ColumnCategory[];
+  columnPickerSettings: string[];
+  onClose: () => void;
+  onDynamicColumnSelect: (columns: string[]) => void;
+}
+
+export const NewColumnPickerCategories: React.FC<
+  NewColumnPickerCategoriesProps
+> = ({
+  columnCategories,
+  onClose,
+  onDynamicColumnSelect,
+  columnPickerSettings,
+}) => {
+  const [selectedCategory, setSelectedCategory] = useState<string | null>(null);
+
+  const currentCategory = useMemo(
+    () => columnCategories.find(category => category.name === selectedCategory),
+    [columnCategories, selectedCategory],
+  );
+
+  const [query, setQuery] = useState<string>("");
+
+  const changeHandler = (column: string) =>
+    columnPickerSettings.includes(column)
+      ? onDynamicColumnSelect(
+          columnPickerSettings.filter(currentCol => currentCol !== column),
+        )
+      : onDynamicColumnSelect([...columnPickerSettings, column]);
+
+  React.useEffect(() => {
+    // Preselect category when there is only one
+    if (columnCategories.length === 1) {
+      setSelectedCategory(columnCategories[0].name);
+    }
+  }, [columnCategories]);
+
+  React.useEffect(() => {
+    if (currentCategory) {
+      setQuery(currentCategory.initialSearch ?? "");
+    }
+  }, [currentCategory]);
+
```

Comment:
suggestions: move this logic to the hooks, one responsibility each:

- selecting category -> one hook with your effect hook etc.
- query -> second hook with your effect hook etc.
- use them here